pb/forms_test.go

```diff
@@ -0,0 +1,357 @@
+	l := &PBLink{Name: &name}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120b1209736f6d65206e616d65",
+		expectedForm: `{
+	"Links": [
+		{
+			"Name": "some name"
+		}
+	]
+}`,
+	})
+}
+
+// same as above but with a Hash
+func TestForm_Links_Hash_some_Name_some(t *testing.T) {
+	name := "some name"
+	l := &PBLink{Hash: cidBytes, Name: &name}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "12160a090155000500010203041209736f6d65206e616d65",
+		expectedForm: `{
+	"Links": [
+		{
+			"Hash": "015500050001020304",
+			"Name": "some name"
+		}
+	]
+}`,
+	})
+}
+
+// this case is excluded from the spec and should not be accepted at
+// higher-level APIs as a Hash property is mandatory
+func TestForm_Links_Tsize_zero(t *testing.T) {
+	var size uint64 = 0
+	l := &PBLink{Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "12021800",
+		expectedForm: `{
+	"Links": [
+		{
+			"Tsize": 0
+		}
+	]
+}`,
+	})
+}
+
+// same as above but with a Hash
+func TestForm_Links_Hash_some_Tsize_zero(t *testing.T) {
+	var size uint64 = 0
+	l := &PBLink{Hash: cidBytes, Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120d0a090155000500010203041800",
+		expectedForm: `{
+	"Links": [
+		{
+			"Hash": "015500050001020304",
+			"Tsize": 0
+		}
+	]
+}`,
+	})
+}
+
+// this case is excluded from the spec and should not be accepted at
+// higher-level APIs as a Hash property is mandatory
+func TestForm_Links_Tsize_some(t *testing.T) {
+	var size uint64 = 1010
+	l := &PBLink{Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120318f207",
+		expectedForm: `{
+	"Links": [
+		{
+			"Tsize": 1010
+		}
+	]
+}`,
+	})
+}
+
+// same as above but with a Hash
+func TestForm_Links_Hash_some_Tsize_some(t *testing.T) {
+	var size uint64 = 1010
+	l := &PBLink{Hash: cidBytes, Tsize: &size}
+	verifyRoundTrip(t, testCase{
+		node:          &PBNode{Links: []*PBLink{l}},
+		expectedBytes: "120e0a0901550005000102030418f207",
+		expectedForm: `{
+	"Links": [
+		{
+			"Hash": "015500050001020304",
+			"Tsize": 1010
+		}
+	]
+}`,
+	})
+}
+
+func verifyRoundTrip(t *testing.T, tc testCase) {
+	actualBytes, actualForm, err := nodeRoundTripToString(tc.node)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if actualBytes != tc.expectedBytes {
+		fmt.Printf(
+			"Expected bytes: [%v]\nGot: [%v]\n",
+			tc.expectedBytes,
+			actualBytes)
+		t.Error("Did not match")
+	}
+
+	if actualForm != tc.expectedForm {
+		fmt.Printf(
+			"Expected form: [%v]\nGot: [%v]\n",
+			tc.expectedForm,
+			actualForm)
+		t.Error("Did not match")
+	}
+}
+
+func nodeRoundTripToString(n *PBNode) (string, string, error) {
+	bytes, err := n.Marshal()
+	if err != nil {
+		return "", "", err
+	}
+	fmt.Printf("[%v]\n", hex.EncodeToString(bytes))
```

Comment:
it would be nicer if this was a `t.Printf`, because then the output is attached to a particular test